### ISSUE TITLE
Investigate KFUNC/KRETFUNC support

### ISSUE THREAD

**Issue by antonblanchard** [linuxppc/issues#363]
A new way of calling into BPF called KFUNC/KRETFUNC was added recently. We should investigate if it will lower the cost of making BPF calls on PowerPC

**Comment by antonblanchard** [comment]
@rnav FYI

**Comment by chleroy** [comment]
Is it https://github.com/linuxppc/linux/commit/e6ac2450d6dee3121cd8bbf2907b78a68a8a353d ?

**Comment by rnav** [comment]
Yes, there are a few items related to BPF that we are working on:
- @BangoriaRavi  is working on adding support for `BPF_PROBE_MEM`, which is required for BTF. I think he is close to posting the patches for this soon. This is required to support kfunc.
- Another area that helps with BPF performance is support for BPF trampolines, as a way to call into BPF programs without going through ftrace. This requires support for ftrace direct calls for which I have posted RFCs and plan to post a v1 soon:
https://patchwork.ozlabs.org/project/linuxppc-dev/list/?series=216968

**Comment by rnav** [comment]
> Is it [linuxppc/linux@e6ac245](https://github.com/linuxppc/linux/commit/e6ac2450d6dee3121cd8bbf2907b78a68a8a353d) ?

Yes. Also: https://lwn.net/Articles/856005/

**Comment by chleroy** [comment]
Maybe porting commit https://github.com/linuxppc/linux/commit/40272035e1d0edcd515ad45be297c4cce044536d to powerpc64 could improve performance ?

**Comment by iamjpn** [comment]
Series porting commit https://github.com/linuxppc/linux/commit/40272035e1d0edcd515ad45be297c4cce044536d to ppc64 posted here: https://patchwork.ozlabs.org/project/linuxppc-dev/list/?series=255361

**Comment by rnav** [comment]
Support for `BPF_PROBE_MEM` has been merged: https://patchwork.ozlabs.org/project/linuxppc-dev/list/?series=284762

Early RFC enabling Trampoline support was posted recently: https://patchwork.ozlabs.org/project/linuxppc-dev/list/?series=284762

**Comment by rnav** [comment]
This is now upstream for ppc64: https://github.com/linuxppc/linux/commit/d243b62b7bd3d5314382d3b54e4992226245e936
ppc32 support pending.
